search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,445 @@
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "ekn_id",
+                                                      G_VARIANT_TYPE_STRING);
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "child_tags",
+                                                      G_VARIANT_TYPE_STRING_ARRAY);
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "tags",
+                                                      G_VARIANT_TYPE_STRING_ARRAY);
+
+      g_variant_builder_close (&builder);
+    }
+
+  shards_strv = strv_from_shard_list ((GSList *) shards);
+
+  eks_article_metadata_complete_query (state->provider->skeleton,
+                                       state->invocation,
+                                       (const gchar * const *) shards_strv,
+                                       g_variant_builder_end (&builder));
+}
+
+static void
+append_construction_prop_from_string (const gchar *key,
+                                      const gchar *str,
+                                      GArray      *values_array,
+                                      GPtrArray   *props_array,
+                                      guint       *nprops)
+{
+  guint index = values_array->len;
+  GValue *value = NULL;
+
+  /* Need to be careful to set value *after* the array has been resized */
+  g_array_set_size (values_array, values_array->len + 1);
+  value = &(g_array_index (values_array, GValue, index));
+  g_value_init (value, G_TYPE_STRING);
+  g_value_set_string (value, str);
+  g_ptr_array_add (props_array, g_strdup (key));
+  ++(*nprops);
+}
+
+static void
+append_construction_prop_from_variant (const gchar *key,
+                                       GVariant    *variant,
+                                       GArray      *values_array,
+                                       GPtrArray   *props_array,
+                                       guint       *nprops)
+{
+  guint index = values_array->len;
+  GValue *value = NULL;
+
+  /* Need to be careful to set value *after* the array has been resized */
+  g_array_set_size (values_array, values_array->len + 1);
+  value = &(g_array_index (values_array, GValue, index));
+  g_dbus_gvariant_to_gvalue (variant, value);
+  g_ptr_array_add (props_array, g_strdup (key));
+  ++(*nprops);
+}
+
+static gchar * valid_query_params[] = {
+  "search-terms",
+  "tags-match-any",
+  "tags-match-all",
+  "limit",
+  "offset"
+};
+static gsize n_valid_query_params = G_N_ELEMENTS (valid_query_params);
+
+static gboolean
+query_parameter_is_valid (const gchar *query_parameter)
```

Comment:
I think we should remove this check; if the key is used as an invalid property, there will be errors when creating the object which will be pretty obvious. I don't see an advantage to doing this check explicitly, and don't like hardcoding the list of properties here like this. (A better check if we really need to do this for some reason is to use `g_object_class_list_properties()` on `EKNC_TYPE_QUERY_OBJECT` instead of hardcoding the list.)

Reply:
Having this check here means that we can return an error to the caller - also, this isn't meant to be a "thin wrapper" over EkncQueryObject. I appreciate that right now the properties are just passed directly to EkncQueryObject right now, however, since we have to maintain compatibility across different SDK versions, it is possible that the API for EkncQueryObject will change over time at which point the properties that EksArticleMetadataProvider and EkncQueryObject will accept will need to diverge. So we need to explicitly check if the caller made a valid request.